Add stretch mode to VideoPlayer control

cart opened the pull request:
Adds stretch mode functionality to the VideoPlayer control.  This is a direct copy from TextureRect, with the exception that I removed STRETCH_SCALE_ON_EXPAND, which appears to be deprecated (but kept around for backward compatibility).

cart committed:
Add stretch mode to VideoPlayer control

akien-mga commented on scene/gui/video_player.h:
```diff
@@ -70,6 +82,7 @@ class VideoPlayer : public Control {
 	int buffering_ms;
 	int audio_track;
 	int bus_index;
+	StretchMode stretch_mode;
```
You should initialize it in the constructor.

cart commented on scene/gui/video_player.cpp:
```diff
@@ -186,8 +186,54 @@ void VideoPlayer::_notification(int p_notification) {
 			if (texture->get_width() == 0)
 				return;
 
-			Size2 s = expand ? get_size() : texture->get_size();
```
Hmm maybe I could re-add STRETCH_SCALE_ON_EXPAND?  Then it would be 100% consistent with TextureRect and fix the api.  Or if we're going to do a hard break and remove expand, shouldn't we remove expand from the other Controls too (TextureRect and TextureButton)?

There are already inconsistencies between the TextureButton and TextureRect implementation (texturebutton only stretches when expand is true).  I'm not sure what the desired final state is.

akien-mga replied:
If we're breaking the API, it would be good to make it consistent for all Controls yes.

If we could find a way that doesn't break the API while adding the feature and making things consistent, that would be great, but I don't really see how. I'll have a closer look at the various Controls with stretch/expand features.

mhilbrunner commented:
Hey! Can you update this to make the changes as commented? Even *if* that means breaking the API, going for consistency is probably worth it.

cart commented:
Sure thing!

cart committed:
Remove expand from texture controls and make stretch modes consistent

cart commented:
Ok I made the changes that I think were expected. All texture controls with stretch_mode now have consistent implementations. Please don't merge until we've discussed everything below :smile: 

This now has the following changes:

* VideoPlayer
    * Copy stretch_mode implementation from TextureRect
    * Removed expand
    * Removed get_minimum_size(), defaulting to Control:get_minimum_size()
* TextureButton
    * Copy stretch_mode implementation from TextureRect
    * Removed expand
    * Removed get_minimum_size(), defaulting to Control:get_minimum_size()
* TextureRect
    * Removed STRETCH_SCALE_ON_EXPAND
    * Set default stretch_mode to STRETCH_SCALE
    * Removed expand
    * Removed get_minimum_size(), defaulting to Control:get_minimum_size()

Removing expand and defaulting to Control:get_minimum_size() means that there is no way to say "set the minimum size to this control's current texture's size". The user must explicitly set a minimum_size value

Consider the following case from project_manager.cpp line 1099 (broken as of this change)
* Node Hierarchy:
    * HboxContainer (recent project row)
        * TextureRect (recent project icon)
* Old behavior:
    * expand defaults to false, stretch_mode is scale_on_expand.  TextureRect's min size is set to the icon's width and height. icon is rendered in the HBoxContainer with it's native width and height and takes up that amount of space
* New behavior (currently broken): 
    * stretch_mode is scale.  TextureRect's min size is (0,0). icon disappears and takes up no space.
    * Fixes:
        1. There are no stretch modes that set min size.  We can change the mode to something like KEEP and the icon will render, but it still takes up no space in the HBoxContainer. The only way to emulate the old behavior is to manually set min_size to (130, 130)
        2. key off of stretch_mode in get_minimum_size().  if stretch_mode is KEEP, set minimum_size to the current texture's size


Also: there is currently a lot of duplicate code across the texture controls.  we should remove that if possible by either:
1. hoisting up the common code into control.cpp (StretchMode enum, and something like stretch_control_to_texture(StretchMode stretch_mode, Ref<Texture> texture)
2. creating a "control stretcher" helper class

reduz commented:
if you want to deprecate the property, just do like this:
```
#ifndef DISABLE_DEPRECATED
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "expand",PROPERTY_HINT_NONE,"",0), "set_expand", "has_expand"); //the zero means no usage, it wont appear in inspector, be serialized, etc.
#endif
```
Then make sure to also mark those methods as deprecated with the same ifdef, so they can be removed when we break compat

akien-mga commented:
I think the new code is a good improvement, this `expand` stuff was pretty convoluted.

We need to preserve compatibility (at least at the API level) with existing projects though if we want to merge it for 3.1, so could you reintroduce `expand` where relevant with its setters/getters, and make it issue a deprecation warning (`WARN_DEPRECATED`) in the setter.

If we find a way to make it convert its expand value to the proper stretch mode value that could be nice, but otherwise a no-op setter that just issues the warning might be enough.

> Consider the following case from project_manager.cpp line 1099 (broken as of this change)

You mean some editor stuff would be broken if we merge this change as is? Would you be able to do a third commit fixing those?

akien-mga commented:
At this stage, I think it's probably best if we merge this for 4.0 and thus don't attempt to preserve compatibility with 3.x. This way we can make a clean and consistent API for this and related controls.

aaronfranke commented:
@cart Is this still desired? If so, it needs to be rebased on the latest master branch.

If not, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

akien-mga commented:
I think this is a good change, it would need to be rebased on the current master branch but from what I remember, it did improve the API consistency and usability.

CC @groud too.

cart commented:
Hey sorry for the delay. This needs more than just a rebase because its a breaking behavior change that effects editor ui (and probably other people's stuff too).

I think a Control refactor that unifies behaviors across controls is extremely desirable, but I'm also currently busy with a bunch of other stuff right now, so I probably won't have the time to do this right.

That being said, unification would be a breaking change, so it is a good candidate for 4.0.

aaronfranke commented:
This PR has not received any new commits for almost 2 years, closing. If this is still desired, it needs to be rebased or recreated.
